rust_src/src/casefiddle.rs

```diff
@@ -36,4 +96,132 @@ pub fn upcase(object: LispObject) -> LispObject {
     unsafe { casify_object(case_action::CASE_UP, object) }
 }
 
+/* Like Fcapitalize but change only the initials.  */
+
+/// Convert the initial of each word in the argument to upper case.
+/// This means that each word's first character is converted to either
+/// title case or upper case, and the rest are left unchanged.  The
+/// argument may be a character or string.  The result has the same
+/// type.  The argument object is not altered--the value is a copy.
+/// If argument is a character, characters which map to multiple code
+/// points when cased, e.g. ﬁ, are returned unchanged.
+#[lisp_fn]
+pub fn upcase_initials(obj: LispObject) -> LispObject {
+    unsafe { casify_object(case_action::CASE_CAPITALIZE_UP, obj) }
+}
+
+// Like Fcapitalize_region but change only the initials.
+
+/// Upcase the initial of each word in the region.
+/// This means that each word's first character is converted to either
+/// title case or upper case, and the rest are left unchanged.  In
+/// programs, give two arguments, the starting and ending character
+/// positions to operate on.
+#[lisp_fn(intspec = "r")]
+pub fn upcase_initials_region(beg: LispObject, end: LispObject) -> LispObject {
+    unsafe { casify_region_nil(case_action::CASE_CAPITALIZE_UP, beg, end) }
+}
+
+/// Convert the region to upper case.  In programs, wants two arguments.
+/// These arguments specify the starting and ending character numbers
+/// of the region to operate on.  When used as a command, the text
+/// between point and the mark is operated on.
+/// See also `capitalize-region'.
+#[lisp_fn(
+    min = "2",
+    intspec = "(list (region-beginning) (region-end) (region-noncontiguous-p))"
+)]
+pub fn upcase_region(
+    beg: LispObject,
+    end: LispObject,
+    region_noncontiguous_p: LispObject,
+) -> LispObject {
+    casefiddle_region(beg, end, region_noncontiguous_p, case_action::CASE_UP)
+}
+
+/// Convert to upper case from point to end of word, moving over.
+///
+/// If point is in the middle of a word, the part of that word before
+/// point is ignored when moving forward.
+///
+/// With negative argument, convert previous words but do not move.
+/// See also `capitalize-word'.
+#[lisp_fn(intspec = "p")]
+pub fn upcase_word(arg: LispObject) -> LispObject {
+    unsafe { casify_word(case_action::CASE_UP, arg) }
+}
+
+// Fiddle with the case of a whole region. Used as a helper by
+// downcase_region and upcase_region.
+fn casefiddle_region(
+    beg: LispObject,
+    end: LispObject,
+    region_noncontiguous_p: LispObject,
+    action: case_action,
+) -> LispObject {
+    if region_noncontiguous_p.is_nil() {
+        unsafe { casify_region_nil(action, beg, end) }
+    } else {
+        let mut bounds = call!(
+            symbol_value(intern("region-extract-function")),
+            intern("bounds")
+        );
+
+        while bounds.is_cons() {
+            let cons = bounds.as_cons_or_error();
```

Comment:
We've just checked that it is a cons, do we need the or_error version here?

Reply:
Maybe something like `while let Some(c) = bounds.as_cons()`.

Reply:
I thought something was ugly about that. Updated.